### PR TITLE
Fix `{f16,f32,f64,f128}::div_euclid`

### PR DESCRIPTION
**Pull request by traviscross**
### DRAFT

We're still analyzing this.

cc #133485

See also:

- https://github.com/rust-lang/rust/issues/133758

"Division and Modulus for Computer Scientists",
Daan Leijen, 2001,
<https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/divmodnote-letter.pdf>

r? ghost

### CODE REVIEW

File: library/std/src/f128.rs [hunk 1]
```diff
@@ -309,8 +309,17 @@ impl f128 {
...
+            return self;
+        }
+        // FIXME(#133755): Though `self % rhs` is documented to be
+        // equivalent to this, it is not, and the distinction matters
+        // here.
+        let r = self - rhs * (self / rhs).trunc();
+        if r < 0.0 {
+            return if rhs > 0.0 { r + rhs } else { r - rhs };
+        }
+        r
```
[the remark this answers is not in the document]
Comment (traviscross) [reply library/std/src/f128.rs:322]: For consistency (both stylistic and as it makes the mathematically analogy more clear), I followed the style of the existing `div_euclid` implementation:

```rust
    pub fn div_euclid(self, rhs: f64) -> f64 {
        let q = (self / rhs).trunc();
        if self % rhs < 0.0 {
            return if rhs > 0.0 { q - 1.0 } else { q + 1.0 };
        }
        q
    }
```

**Comment by traviscross** [comment]
In talking with @ehuss, he pointed out [RFC 2169](https://github.com/rust-lang/rfcs/pull/2169) where `div_euclid` and `rem_euclid` were proposed.

In looking at that, interestingly, the definition given for `div_euclid` matches exactly what's in the standard library today.  And more interestingly, the definition that it gives for `rem_euclid`...

```rust
fn mod_euc(self, rhs: f64) -> f64 {
    let r = self % rhs;
    if r < 0.0 {
        return if rhs > 0.0 { r + rhs } else { r - rhs }
    }
    r
}
```

...matches exactly the original implementation in this PR (which, due to #133758, is now being adjusted slightly).

I wonder what the history was in moving from the definition in the RFC, which seems more obvious and is in line with the 2001 Leijen paper, to the one now in the standard library.

**Comment by traviscross** [comment]
Looking at the git history, the current peculiar implementation goes all the way back to the original PR:

- https://github.com/rust-lang/rust/pull/49389

Looking through the discussion there, and also in the RFC thread and the tracking issue...

- https://github.com/rust-lang/rfcs/pull/2169
- https://github.com/rust-lang/rust/issues/49048

...I don't immediately see any discussion about the mathematics to justify the change to the current implementation, which is:

```rust
pub fn rem_euclid(self, rhs: f64) -> f64 {
    let r = self % rhs;
    if r < 0.0 { r + rhs.abs() } else { r }
}
```

**Comment by quaternic** [comment]
Because,
 - `r - rhs` is equivalent to `r + (-rhs)`
 - `rhs.abs()` is equivalent to `if rhs.is_sign_positive() { rhs } else { -rhs }`
 
So the changed expression would only differ when `rhs` is `+NaN` or `0.0`, which aren't possible within the `r < 0.0` case. That is, the two implementations are exactly equivalent, I believe.

**Comment by tczajka** [comment]
This PR should be rejected. `rem_euclid` is correct, it's `div_euclid` that is incorrect, as discussed in #107904.

**Comment by traviscross** [comment]
It's a draft, so there's nothing to reject at this point.  And yes, I agree about that.

**Comment by traviscross** [comment]
(The state of this PR lags the current state of my analysis.  I have a hypothesis for a fix for `div_euclid` that I'm currently testing.)

**Comment by traviscross** [comment]
Closing in favor of:

- https://github.com/rust-lang/rust/pull/134062

(...so as to fix the branch name.)

**Pull request closed without merging**